src/containers/EditListingPage/EditListingWizard/EditListingDetailsPanel/EditListingDetailsPanel.js

```diff
@@ -142,6 +144,38 @@ const initialValuesForListingFields = (
   }, {});
 };
 
+// Generates initial values for listing categories based on provided public data and configuration.
+// This function validates if the initial values match with the configuration received via assets.
+// If a categoryLevel value doesn't match with the category configuration, it is not passed on to the form.
+const initialValuesForListingCategories = (publicData, listingCategoriesConfig) => {
+  const categories = {};
+  let level = 1;
+  let currentCategoryLevel = listingCategoriesConfig;
+
+  while (true) {
+    const currentCategoryKey = `${CATEGORY_PREFIX}${level}`;
+    const currentCategoryValue = publicData[currentCategoryKey];
+
+    if (!currentCategoryValue) break;
+
+    const found = currentCategoryLevel.find(category => category.id === currentCategoryValue);
+
+    if (!found) break;
+
+    categories[currentCategoryKey] = publicData[currentCategoryKey];
+
+    if (publicData[`${CATEGORY_PREFIX}${level + 1}`] && found.subcategories) {
+      currentCategoryLevel = found.subcategories;
+    } else {
+      break;
+    }
+
+    level++;
+  }
+
+  return categories;
+};
```

Comment:
Another way to write this is to go a bit more towards functional programming, which makes things more declarative and readable compared to a while loop with carefully crafted breaks and if-else statements.

```js
const pickCategoryFields = (data, level, categoryLevelOptions=[]) => {
  const currentCategoryKey = `${CATEGORY_PREFIX}${level}`;
  const currentCategoryValue = data[currentCategoryKey];
  const isCategoryLevelSet = typeof currentCategoryValue !== 'undefined';

  // Validate the value against category options
  const categoryOptionConfig = categoryLevelOptions.find(category => category.id === currentCategoryValue);
  const isValidCategoryValue = !!categoryOptionConfig;
  const nextLevelOptions = categoryOptionConfig?.subcategories || [];

  // Return category level property if it's found from the data and the value is one of the valid options.
  // Go through all the nested levels.
  return isCategoryLevelSet && isValidCategoryValue
    ? { [currentCategoryKey]: currentCategoryValue, ...pickCategoryFields(data, ++level, nextLevelOptions) }
    : {};
}
```

```js
const initialValuesForListingCategories = pickCategoryFields(publicData, 1, listingCategoriesConfig);
```